File: ai-tools-engine/engine/src/main/kotlin/cz/cleanship/aitools/engine/tools/PromptPrinter.kt
package cz.cleanship.aitools.engine.tools

import cz.cleanship.aitools.engine.io.Output
import cz.cleanship.aitools.engine.models.FragmentManifest
import cz.cleanship.aitools.engine.models.PromptManifest
import cz.cleanship.aitools.engine.models.RulesetManifest

class PromptPrinter(
    private val rulesetResolver: RulesetResolver = RulesetResolver(),
    private val fragmentResolver: FragmentResolver = FragmentResolver(),
) : Printer<PromptContext> {

    override fun print(entity: PromptContext, output: Output): Output {
        val (prompt, rulesets, allRulesets, fragments, allFragments) = entity
        output.appendTextTopic("# ${prompt.id}", prompt.description)

        output.appendListTopic(
            "## Variables",
            prompt.variables.map {
                val required = if (it.required) " (required)" else ""
                "`{{${it.name}}}`$required: ${it.description}"
            },
        )

        val matchedRulesets = rulesetResolver.resolve(
            patterns = prompt.rulesets,
            available = rulesets,
            requestedBy = "prompt '${prompt.id}'",
            allRulesets = allRulesets,
        )
        output.appendListTopic("## Rules", matchedRulesets.flatMap { it.rules } + prompt.rules)

        output.appendTextTopic("## Prompt", entity.prompt.content)

        val matchedFragments = fragmentResolver.resolve(
            patterns = prompt.fragments,
            available = fragments,
            requestedBy = "prompt '${prompt.id}'",
            allFragments = allFragments,
        )
        if (matchedFragments.isNotEmpty()) {
            output.appendLine("## Fragments")
            output.appendLine()
            for (fragment in matchedFragments) {
                output.appendTextTopic("### ${fragment.id}", fragment.content)
            }
        }

        return output
    }
}

data class PromptContext(
    val prompt: PromptManifest,
    val rulesets: Map<String, RulesetManifest>,
    val allRulesets: Map<String, RulesetManifest> = rulesets,
    val fragments: Map<String, FragmentManifest> = emptyMap(),
    val allFragments: Map<String, FragmentManifest> = fragments,
)

File: src/tools/common/promptPrinter.ts
import { Prompt } from '../../core/models/types';
import { FileContent } from './fileContent';
import { Printer } from './printer';

export class PromptPrinter implements Printer<Prompt> {
  print(entity: Prompt, content: FileContent): FileContent {
    content.append(`# ${entity.id}`);
    content.append('');
    content.append(entity.description);
    content.append('');

    if (entity.variables && entity.variables.length > 0) {
      content.append('## Variables');
      content.append('');
      for (const variable of entity.variables) {
        const required = variable.required ? ' (required)' : '';
        content.append(`- \`{{${variable.name}}}\`${required}: ${variable.description || ''}`);
      }
      content.append('');
    }

    if (entity.content) {
      content.append('## Prompt');
      content.append('');
      content.append(entity.content);
      content.append('');
    }

    if (entity.system) {
      content.append('## System Prompt');
      content.append('');
      content.append(entity.system);
      content.append('');
    }

    if (entity.user) {
      content.append('## User Prompt');
      content.append('');
      content.append(entity.user);
      content.append('');
    }
    return content;
  }
}
